addons/base_import/models.py

```diff
@@ -126,6 +153,65 @@ def get_fields(self, cr, uid, model, context=None,
         # TODO: cache on model?
         return fields
 
+    def _file_type_handler(self, file_type, record, options):
+        file_extension = FILE_TYPE_DICT.get(file_type)
+        if file_extension:
+            return getattr(self, '_read_' + file_extension)(record, options)
+        else:
+            raise UserError(_("Not a compatible file format. It should be CSV, XLS, XLSX or ODS file"))
+
+    def _read_xls(self, record, options):
+        if not xlrd:
+            raise UserError(_(" XLRD library not found. XLS/XLSX file would not import."))
+        if xlrd.__VERSION__.split('.') < ['0', '9', '3']:
+            raise UserError(_(" Lower Version of XLRD, You need to install latest version of xlrd."))
+        book = xlrd.open_workbook(file_contents=record.file)
+        bk = book.sheet_by_index(0)
+        for row_index in range(bk.nrows):
+            keys = []
+            for col_index in range(bk.ncols):
+                if bk.cell_type(row_index, col_index) == xlrd.XL_CELL_NUMBER:
+                    number_decimal = bk.cell_value(row_index, col_index) % 1
+                    if number_decimal == 0.0:
+                        keys.append(str(int(bk.cell_value(row_index, col_index))))
+                    else:
+                        keys.append(str(bk.cell_value(row_index, col_index)))
+                elif bk.cell_type(row_index, col_index) == xlrd.XL_CELL_DATE:
+                    date_decimal = bk.cell_value(row_index, col_index) % 1
+                    if date_decimal == 0.0:
+                        keys.append(xlrd.xldate.xldate_as_datetime(bk.cell_value(row_index, col_index), book.datemode).strftime('%Y-%m-%d'))
+                    else:
+                        keys.append(xlrd.xldate.xldate_as_datetime(bk.cell_value(row_index, col_index), book.datemode).strftime('%Y-%m-%d %H:%M:%S'))
+                elif bk.cell_type(row_index, col_index) == xlrd.XL_CELL_BOOLEAN:
+                    if bk.cell_value(row_index, col_index) == 1:
+                        keys.append('True')
+                    else:
+                        keys.append('False')
+                elif bk.cell_type(row_index, col_index) == xlrd.XL_CELL_ERROR:
+                    if bk.cell_value(row_index, col_index) in xlrd.error_text_from_code:
+                        raise UserError(_("Error during reading XLS/XLSX file. Error Code: %s"% xlrd.error_text_from_code[bk.cell_value(row_index, col_index)] ))
+                else:
+                    keys.append(bk.cell_value(row_index, col_index))
+            yield keys
+
+    _read_xlsx = _read_xls
+
+    def _read_ods(self, record, options):
+        if not odf:
+            raise UserError(_(" ODFPY library not found. ODS file would not import."))
+        f = io.BytesIO()
+        f.write(base64.decodestring(base64.b64encode(record.file)))
```

Comment:
What is the point of encoding file data to base64 and immediately decoding it back to binary?